### PR TITLE
Check types with mypy

### LINKED ISSUE #56: gradual type checking with mypy
It'd be nice to have some type checking. I just made a [silly mistake](https://github.com/goavki/apertium-apy/commit/988fb1ae9fc2fb1a20acb4e00cf1053c27fee3d6) when renaming a variable (sending the *module* named html into re.sub instead of the html-contents variable …), which a good type checker would've caught right away. Python can't do very good type checking, but at least *some* type checking is possible with mypy, and `mypy --check-untyped-defs` would've caught this.

We'd have to do any type annotations in comments instead of the nice inline [PEP484](https://www.python.org/dev/peps/pep-0484/) syntax, since we want to support older Python versions.

We might have to only enable it for newer python versions in travis.

We might have to include some stubs for modules (`servlet.py:55: error: Cannot find module named 'cld2full'`), or tell mypy to ignore some things (`translation.py:10: error: Name 'locks' already defined` because we fallback to toro.locks in a try-except) or fill in for some bugs in mypy (`translation.py:12: error: Module 'select' has no attribute 'PIPE_BUF'`).

### PR DESCRIPTION
**Pull request by Androbin**
Fixes #56
@shardulc @sushain97 Friendly ping.
@unhammer FYI

### CODE REVIEW

**Comment by sushain97** [comment]
Doesn't this break the build? I don't consider #56 fixed until it doesn't. Preferably also without grepping out errors.

**Comment by Androbin** [comment]
> Doesn't this break the build?

Okay, I'll try to fix the reported errors.

> Preferably also without grepping out errors.

I wanted to filter the non-error noise.
But maybe I can get rid of these warnings as well.

**Comment by Androbin** [comment]
@sushain97 I have fixes all reported errors.
The grepping removes notes:
`gateway.py:15: note: (Stub files are from https://github.com/python/typeshed)`
and reports for missing stub files:
```
gateway.py:15: error: No library stub file for module 'tornado'
gateway.py:16: error: No library stub file for module 'tornado.httpserver'
gateway.py:17: error: No library stub file for module 'tornado.web'
gateway.py:18: error: No library stub file for module 'tornado.httpclient'
gateway.py:21: error: No library stub file for module 'tornado.log'
language-names/scraper.py:9: error: No library stub file for module 'lxml'
servlet.py:26: error: No library stub file for module 'tornado.locks'
servlet.py:29: error: No library stub file for module 'tornado'
servlet.py:30: error: No library stub file for module 'tornado.web'
servlet.py:31: error: No library stub file for module 'tornado.httpserver'
servlet.py:32: error: No library stub file for module 'tornado.httputil'
servlet.py:33: error: No library stub file for module 'tornado.process'
servlet.py:34: error: No library stub file for module 'tornado.iostream'
servlet.py:38: error: No library stub file for module 'tornado.escape'
servlet.py:40: error: No library stub file for module 'tornado.log'
servlet.py:67: error: No library stub file for module 'chardet'
toro.py:25: error: No library stub file for module 'tornado'
toro.py:27: error: No library stub file for module 'tornado.concurrent'
translation.py:5: error: No library stub file for module 'tornado'
translation.py:6: error: No library stub file for module 'tornado.process'
translation.py:7: error: No library stub file for module 'tornado.iostream'
translation.py:9: error: No library stub file for module 'tornado.locks'
translation_py32.py:5: error: No library stub file for module 'tornado'
translation_py32.py:6: error: No library stub file for module 'tornado.process'
translation_py32.py:7: error: No library stub file for module 'tornado.iostream'
translation_py32.py:9: error: No library stub file for module 'tornado.locks'
```

If we find such, we can add them later.

**Comment by sushain97** [comment]
Does mypy have no command line flags to turn off notes? Also, looks like the build is very broken now.

**Comment by Androbin** [comment]
> Does mypy have no command line flags to turn off notes?

No, unfortunately not.

> Also, looks like the build is very broken now.

Had to fix some `flake8` errors.

**Comment by sushain97** [comment]
Still failing?

**Comment by Androbin** [comment]
There seems to be some non-related issue with the pip package:
https://travis-ci.org/goavki/apertium-apy/jobs/322248253#L512

File: gateway.py [hunk 1]
```diff
@@ -10,7 +10,8 @@
 import socket
 import servlet
 import pprint
-from collections import OrderedDict, defaultdict
+from typing import Any, Dict, List, Set, Tuple  # noqa
```
Comment (sushain97) [inline gateway.py]: Please specify exactly which error in this places so that we don't potentially silence useful errors.

Comment (Androbin) [reply gateway.py]: * `typing` imports are only ever used by `mypy` and are unused as far as `flake8` is concerned.
* `try: import foo except: import foobar as foo` is detected as re-assignment.
* Certain modules are just not recognized by `mypy`.

Comment (sushain97) [reply gateway.py]: no, I don't mean like that. the # noqa syntax lets you specify which error you're ignoring instead of all of them

**Comment by sushain97** [comment]
Huh, I restarted the build and it still seems to be failing?

**Comment by Androbin** [comment]
Can't figure it out either...

**Comment by Androbin** [comment]
@sushain97 I figured it out!
* `3.2` failed because `mypy` isn't available below `3.3` to install or import
* All tests failed because of the non-existent `translation.CatPipeline`
* Also, I refactored the Travis commands

**Comment by sushain97** [comment]
Ugh, this is yuck. @unhammer how much do we want to support py3.2? Also asked Tino in IRC.

**Comment by Androbin** [comment]
@shardulc FYI. (You wanted to be notified of `apy` PR's.)

**Comment by Androbin** [comment]
Can we sort out the py3.2 issue elsewhere?
What is left to do for this PR?

**Comment by sushain97** [comment]
Counted as part of GCI task https://codein.withgoogle.com/dashboard/task-instances/4764436512374784/

**Pull request merged**